### ISSUE TITLE
React Native implementation

### ISSUE THREAD

**Issue by mrousavy** [woltapp/blurhash#54]
### UPDATE: I've created a native UI module for iOS (Swift) and Android (Kotlin) myself, check it out here: [mrousavy/react-native-blurhash](https://github.com/mrousavy/react-native-blurhash)

Hi! The library is really awesome, but I can't seem to find a way to use it in **React Native** (with the _blurhash-typescript_ library).
The `decode(...)` function returns a raw pixel array (`UInt8ClampedArray`), but the `<Image>` component can't be initialized with a pixel array. (see: [react-native Image.source](https://reactnative.dev/docs/image#source))

**Comment by espipj** [comment]
I'm happy to help with this, I've been trying by using react-native-canvas but it is not performant at all...
It would be amazing a decodeToBase64 method, it would be quite easy from there to give support for React Native... 📱

**Comment by mrousavy** [comment]
@espipj yes, I've thought about using the native APIs aswell by implementing this extra initializing method in the existing, already very fast, fast-image component library. [I've also created an issue over there](https://github.com/DylanVann/react-native-fast-image/issues/669)

Unfortunately I don't have the time right now to create such a pull request, from what I understood though it doesn't seem that hard when I looked into the native code of fast-image.

**Comment by espipj** [comment]
Hey @mrousavy I got something working on iOS using the swift implementation and linking it to React Native
<img width="512" alt="Screenshot 2020-05-13 at 23 56 31" src="https://user-images.githubusercontent.com/12865914/81874301-85662e80-9575-11ea-885f-467d767e0b61.png">
I'm working on it at https://github.com/espipj/react-native-blurhash
I might need some help as my Swift and Kotlin experience is none 
cc/ @LuisRodriguezLD

**Comment by schiller-manuel** [comment]
@mrousavy I implemented blurhash using OpenGL so that it can be used in React Native.
The package is available here: https://github.com/schiller-manuel/gl-react-blurhash

**Comment by mrousavy** [comment]
@schiller-manuel that looks really interesting! your idea of using this would be to render this GL View as a placeholder until the image (which is a separate, invisible component) has been loaded, to then swap the GL View with the Image view, correct?

**Comment by schiller-manuel** [comment]
yes, that's how I'd use it. In the [linked example](https://snack.expo.io/@git/github.com/schiller-manuel/gl-react-blurhash:example@develop) the `onPress` event triggers the swapping of placeholder and image, but it could also be the `onLoad` event of the actual image.

**Comment by mrousavy** [comment]
@espipj @schiller-manuel I've just created a native UI module for this. See: https://github.com/mrousavy/react-native-blurhash.

<img src="https://user-images.githubusercontent.com/15199031/84646392-517a8200-af02-11ea-8051-e96d90790a02.png" width=300>



Currently, I've only implemented iOS, working on Android right now. It uses the default Image component, I still need to create extra properties like `resizeMethod`, `resizeMode`, and maybe `borderRadius`. If you could help me out on that and create some PRs, I'd gladly accept them.

In short: It's a lightweight component that just renders the Blurhash Image in react native. No need for `react-native-unimodules`, `gl-react`, `gl-react-native`, `buffer`, ...

**Comment by mrousavy** [comment]
I did it. It's working on Android and iOS. Uses the native Image components, and no dependencies.

Here's the link: https://github.com/mrousavy/react-native-blurhash
It's public on npm: https://www.npmjs.com/package/react-native-blurhash

@woltapp do you want to mention this library in the README in case anyone else is searching for a React Native implementation?

**Comment by siddhatiwari** [comment]
Is there a react-native library that doesn't require native linking?

**Comment by schiller-manuel** [comment]
@siddhatiwari you can use [gl-react-blurhash](https://github.com/schiller-manuel/gl-react-blurhash) in Expo if this is what you want to do.

**Comment by mrousavy** [comment]
@siddhatiwari while the expo library works, you should keep in mind that the performance is not as good as my native module, since it uses the JS thread for decoding, which is slower and blocks any other JS code execution. @schiller-manuel did an awesome job implementing this without native code, but it's just limited performance wise per design. Always choose the right tool for the job.

**Comment by siddhatiwari** [comment]
@mrousavy @schiller-manuel Thanks for the quick response! Will check out gl-react-blurhash.

**Comment by jerry-git** [comment]
> @woltapp do you want to mention this library in the README in case anyone else is searching for a React Native implementation?

Hi @mrousavy! Firstly, sorry for the super late reply. Secondly, yes let's have it documented in the readme. Would you mind providing a PR for that? 🙂

**Comment by mrousavy** [comment]
@jerry-git sure, here you go: https://github.com/woltapp/blurhash/pull/176

**Comment by Thisen** [comment]
Closing, as this is now third-party :)
